[Site Isolation] Isolated subframes should use the user agent set from website policies

charliewolfe opened the pull request:
#### 4aa7c4c9b5a35bf5e68871fd6ff955ced5011869
<pre>
[Site Isolation] Isolated subframes should use the user agent set from website policies
<a href="https://bugs.webkit.org/show_bug.cgi?id=269213">https://bugs.webkit.org/show_bug.cgi?id=269213</a>
<a href="https://rdar.apple.com/122817201">rdar://122817201</a>

Reviewed by Alex Christensen.

To ensure that the correct website policies are sent to navigating subframes, we need to store the
website policies when a main frame provisional load has started. The page should not be updated with the
stored website policies until the provisional load has been committed.

The tests verify that the subframe has the correct user agent and that the website policies are only
applied once a provisional load has been committed.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::userAgent const):
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::customUserAgent const):
(WebCore::LocalFrame::customUserAgentAsSiteSpecificQuirks const):
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::customUserAgent const):
(WebCore::RemoteFrame::customUserAgentAsSiteSpecificQuirks const):
* Source/WebCore/page/RemoteFrame.h:
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
(WebKit::ProvisionalPageProxy::mainFrameWebsitePoliciesData const):
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::injectPageIntoNewProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::commitProvisionalPage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::applyWebsitePolicies):
(WebKit::WebLocalFrameLoaderClient::applyToDocumentLoader): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::applyWebsitePolicies):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::remoteFrameClient const):
(WebKit::WebFrame::didReceivePolicyDecision):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_unifiedTextReplacementController):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/274779@main">https://commits.webkit.org/274779@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3834c9266c6b542dc1e97a0354d9a7cd28bae376

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39994 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19005 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42370 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42539 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42301 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21913 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16335 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40568 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16006 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/13854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/35510 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43817 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/36294 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/35850 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/12170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/16428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/16477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5278 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1311,6 +1311,20 @@ void LocalFrame::didAccessWindowProxyPropertyViaOpener(WindowProxyProperty prope
+{
+    if (RefPtr documentLoader = loader().activeDocumentLoader())
+        return documentLoader->customUserAgent();
+    return ""_s;
```
I feel like this should be the null string rather than the empty string.

achristensen07 commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1311,6 +1311,20 @@ void LocalFrame::didAccessWindowProxyPropertyViaOpener(WindowProxyProperty prope
+{
+    if (RefPtr documentLoader = loader().activeDocumentLoader())
+        return documentLoader->customUserAgentAsSiteSpecificQuirks();
+    return ""_s;
```
ditto.

achristensen07 commented on Source/WebKit/UIProcess/ProvisionalPageProxy.h:
```diff
@@ -206,6 +209,7 @@ class ProvisionalPageProxy : public IPC::MessageReceiver, public IPC::MessageSen
     bool m_needsDidStartProvisionalLoad { true };
     URL m_provisionalLoadURL;
     WebPageProxyMessageReceiverRegistration m_messageReceiverRegistration;
+    std::optional<WebsitePoliciesData> m_websitePoliciesData;
```
"mainFrame" might be good to have here.  Same with setter and getter.

achristensen07 commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -424,7 +424,8 @@ void WebFrameProxy::prepareForProvisionalNavigationInProcess(WebProcessProxy& pr
         LocalFrameCreationParameters localFrameCreationParameters {
-            m_provisionalFrame->layerHostingContextIdentifier()
+            m_provisionalFrame->layerHostingContextIdentifier(),
+            m_page->websitePoliciesData()
```
I think this should only be present for the main frame, which is never transitioned to local, so this should be nullopt.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5863,6 +5865,13 @@ void WebPageProxy::didStartProvisionalLoadForFrameShared(Ref<WebProcessProxy>&&
         internals().pageLoadState.didStartProvisionalLoad(transaction, url.string(), unreachableURL.string());
         protectedPageClient->didStartProvisionalLoadForMainFrame();
         closeOverlayedViews();
+        if (auto* websitePolicies = navigation->websitePolicies()) {
```
I don't think this is the right place for this logic.  The lambda in WebPageProxy::decidePolicyForNavigationAction receives them from the client and shouldn't put them on the navigation.  It should be retained with a RefPtr and passed to WebPageProxy::receivedNavigationActionPolicyDecision as a separate parameter.  That'll take some untangling that should probably be done in a separate PR.  Then in WebPageProxy::receivedNavigationActionPolicyDecision we should store them on the ProvisionalPageProxy and when committing a ProvisionalPageProxy we should move them from the ProvisionalPageProxy to the WebPageProxy.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -3471,6 +3474,8 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     RefPtr<WebPageProxy> m_pageToCloneSessionStorageFrom;
     Ref<BrowsingContextGroup> m_browsingContextGroup;
+
+    std::optional<WebsitePoliciesData> m_websitePoliciesData;
```
"mainFrame" should be in name.

achristensen07 commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1311,6 +1311,20 @@ void LocalFrame::didAccessWindowProxyPropertyViaOpener(WindowProxyProperty prope
+{
+    if (RefPtr documentLoader = loader().activeDocumentLoader())
+        return documentLoader->customUserAgent();
+    return nullString();
```
This is really only needed when you return a const String&.  Here, you return a String, so you can just return { }.  Ditto below.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -4345,6 +4345,9 @@ void WebPageProxy::receivedNavigationActionPolicyDecision(WebProcessProxy& proce
         }
 
+        if (auto* websitePolicies = navigation->websitePolicies(); websitePolicies && frame->isMainFrame())
+            m_mainFrameWebsitePoliciesData = websitePolicies->data();
```
This is not the correct time to set this.  If you're on a.com and navigating to b.com, there is still a lot of time between this point and when the navigation is committed when we can make an iframe, and that iframe would have b.com's policies.  It looks like you already have the right code in WebPageProxy::commitProvisionalPage, so this should be removed.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2372,6 +2372,8 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 
     void addConsoleMessage(WebCore::FrameIdentifier, JSC::MessageSource, JSC::MessageLevel, const String&, std::optional<WebCore::ResourceLoaderIdentifier> = std::nullopt);
 
+    std::optional<WebsitePoliciesData> mainFrameWebsitePoliciesData() const { return m_mainFrameWebsitePoliciesData; }
```
const std::optional<WebsitePoliciesData>&
Avoid unnecessary copies.
Also see below about making this a unique_ptr.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -3471,6 +3473,8 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     RefPtr<WebPageProxy> m_pageToCloneSessionStorageFrom;
     Ref<BrowsingContextGroup> m_browsingContextGroup;
+
+    std::optional<WebsitePoliciesData> m_mainFrameWebsitePoliciesData;
```
Let's make this a unique_ptr.  Then we don't need to include WebsitePoliciesData.h.  If we really do want to use std::optional, then it should go in WebPageProxyInternals.h.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -28,6 +28,7 @@
 #include "APIObject.h"
 #include "MessageReceiver.h"
 #include "MessageSender.h"
+#include "WebsitePoliciesData.h"
```
Don't add more includes here.  This is widely included and we've worked hard to reduce the includes here to reduce compile time.

charliewolfe committed:
[Site Isolation] Isolated subframes should use the user agent set from website policies
https://bugs.webkit.org/show_bug.cgi?id=269213
rdar://122817201

Reviewed by Alex Christensen.

To ensure that the correct website policies are sent to navigating subframes, we need to store the
website policies when a main frame provisional load has started. The page should not be updated with the
stored website policies until the provisional load has been committed.

The tests verify that the subframe has the correct user agent and that the website policies are only
applied once a provisional load has been committed.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::userAgent const):
* Source/WebCore/page/Frame.h:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::customUserAgent const):
(WebCore::LocalFrame::customUserAgentAsSiteSpecificQuirks const):
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::customUserAgent const):
(WebCore::RemoteFrame::customUserAgentAsSiteSpecificQuirks const):
* Source/WebCore/page/RemoteFrame.h:
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
(WebKit::ProvisionalPageProxy::mainFrameWebsitePoliciesData const):
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::injectPageIntoNewProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision):
(WebKit::WebPageProxy::commitProvisionalPage):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::applyWebsitePolicies):
(WebKit::WebLocalFrameLoaderClient::applyToDocumentLoader): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::applyWebsitePolicies):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::remoteFrameClient const):
(WebKit::WebFrame::didReceivePolicyDecision):
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_unifiedTextReplacementController):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/274779@main
